core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -480,8 +482,10 @@ protected Scope visitInsert(Insert insert, Optional<Scope> scope)
                     "INSERT",
                     targetTable,
                     Optional.empty(),
-                    Optional.of(insertColumns.stream()
-                            .map(insertColumn -> new Column(insertColumn, tableMetadata.getColumn(insertColumn).getType().toString()))
+                    Optional.of(Streams.zip(
+                            insertColumns.stream(),
+                            queryScope.getRelationType().getVisibleFields().stream(),
+                            this::createOutputColumn)
```

Comment:
The new method uses field.getType() for the Column type, instead of tableMetadata.getColumn(insertColumn).getType(). The types of are necessarily the same (see typesMatchForInsert)

Reply:
https://github.com/trinodb/trino/pull/7465#discussion_r606285522